### PR TITLE
plan: refactor the code of building `Insert`.

### LINKED ISSUE #7061: Wrong behavior when insert default value
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table t(a int default 1, b int default 2);
insert into t (b) values(default);
```

2. What did you expect to see?

One row which a is 1, b is 2.

3. What did you see instead?

One row which a is 1, b is 1.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

cf8a4129a5d76a29e34cd1f4c7a6514b60ab1ea4

### LINKED ISSUE #7064: Error when executing `insert select` where table has generated column
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table t(a int, b int as (a+1), c int);
insert into t (a) select a from t;
insert into t (a, c) select a, c from t;
```

2. What did you expect to see?

No error.

3. What did you see instead?

Error returned.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

cf8a4129a5d76a29e34cd1f4c7a6514b60ab1ea4

### PR DESCRIPTION
**Pull request by winoros**
## What have you changed? (mandatory)

Refactor the code when building `Insert` plan.
To fix some bugs make code more maintainable.
Fix #7064 
Fix #7061 
And another bug:
the table have 3 columns.
`insert into t values(default, default, default, default)` will make tidb panicked.

Main change:
We used to check the length of value_list and column_list when executing. Now we move this change to planner.

## What is the type of the changes? (mandatory)

- Refactor
- Improvement
- Bug fix

## How has this PR been tested? (mandatory)

existing tests and newly added tests.

## Does this PR need to be added to the release notes? (mandatory)

Fix the behavior of `INSERT` statement in some scenarios.

### CODE REVIEW

**Comment by jackysp** [comment]
Please fix CI @winoros

**Comment by winoros** [comment]
Integration-common-test is failed because the bug in tikv.
Now it's ready to review.

File: plan/planbuilder.go [hunk 1]
```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
 		}
+		onDupColSet[column.Name.L] = struct{}{}
+		dupCols = append(dupCols, col)
 	}
+	return onDupColSet, dupCols, nil
+}
+
+func (p *Insert) getAffectCols(insert *ast.InsertStmt) (affectedValuesCols []*table.Column, err error) {
+	if len(insert.Columns) > 0 {
+		// Branch for `INSERT INTO tbl_name (col_name [, col_name] ...) {VALUES | VALUE} (value_list) [, (value_list)] ...`,
```
Comment (zz-jason) [inline plan/planbuilder.go]: How about:
```go
// This is for the following scenario:
// 1. `INSERT INTO tbl_name (col_name [, col_name] ...) {VALUES | VALUE} ...`
// 2. `INSERT INTO tbl_name (col_name [, col_name] ...) SELECT ...`
```

File: plan/planbuilder.go [hunk 2]
```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
+		if column.IsGenerated() {
+			return nil, nil, ErrBadGeneratedColumn.GenByArgs(assign.Column.Name.O, tblInfo.Name.O)
 		}
+		onDupColSet[column.Name.L] = struct{}{}
+		dupCols = append(dupCols, col)
 	}
+	return onDupColSet, dupCols, nil
+}
+
+func (p *Insert) getAffectCols(insert *ast.InsertStmt) (affectedValuesCols []*table.Column, err error) {
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply plan/planbuilder.go]: how about s/getAffectCols/getInsertColumns/?

File: plan/planbuilder.go [hunk 3]
```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
+		// No generated column is allowed.
+		for _, col := range affectedValuesCols {
 			if col.IsGenerated() {
-				b.err = ErrBadGeneratedColumn.GenByArgs(col.Name.O, tableInfo.Name.O)
-				return nil
+				b.err = ErrBadGeneratedColumn.GenByArgs(col.Name.O, insertPlan.Table.Meta().Name.O)
+				return
 			}
 		}
+		// If the value_list and col_list is empty and we have generated column, we can still write to this table.
```
Comment (zz-jason) [inline plan/planbuilder.go]: This comment should reside to `if len(insert.Columns) == 0 && len(insert.Lists[0]) == 0` branch? Maybe we can remove it?

File: plan/planbuilder.go [hunk 4]
```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
-		if b.err != nil {
-			return nil
+	totalTableCols := insertPlan.Table.Cols()
+	for i, valuesItem := range insert.Lists {
+		// The length of the value_list should keep unchanged.
+		// "insert into t values (), ()" is valid.
+		// "insert into t values (), (1)" is not valid.
+		// "insert into t values (1), ()" is not valid.
+		// "insert into t values (1,2), (1)" is not valid.
+		if i > 0 && len(insert.Lists[i-1]) != len(valuesItem) {
```
Comment (zz-jason) [inline plan/planbuilder.go]: Maybe this form can help reading the code:
```go
if i > 0 && len(insert.Lists[i]) != len(insert.Lists[i-1])
```

File: plan/planbuilder.go [hunk 5]
```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
+			b.err = ErrWrongValueCountOnRow.GenByArgs(i + 1)
+			return
 		}
-
-		insertPlan.Schema4OnDuplicate = expression.MergeSchema(insertPlan.tableSchema, insertPlan.SelectPlan.Schema())
+		exprList := make([]expression.Expression, 0, len(valuesItem))
+		for j, valueItem := range valuesItem {
+			var expr expression.Expression
+			var err error
+			if dft, ok := valueItem.(*ast.DefaultExpr); ok {
```
Comment (zz-jason) [inline plan/planbuilder.go]: Usually the abbreviation for `default` is `def`, see https://www.abbreviations.com/abbreviation/Default

Comment (zz-jason) [inline plan/planbuilder.go]: I think the "type switch" is more suitable.

File: plan/planbuilder.go [hunk 6]
```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
+		// If the value_list and col_list is empty and we have generated column, we can still write to this table.
+		// i.e. insert into t values(); can be executed successfully if t have generated column.
+	}
 
-		insertPlan.SelectPlan, b.err = doOptimize(b.optFlag, selectPlan.(LogicalPlan))
-		if b.err != nil {
-			return nil
+	totalTableCols := insertPlan.Table.Cols()
+	for i, valuesItem := range insert.Lists {
+		// The length of the value_list should keep unchanged.
```
Comment (zz-jason) [inline plan/planbuilder.go]: How about:
```go
// The lengths of all the value lists should be the same, e.g:
```

File: plan/planbuilder.go [hunk 7]
```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
+		b.err = ErrWrongValueCountOnRow.GenByArgs(1)
+		return
+	}
 
-		// Construct the function which calculates the assign value of the column.
-		expr, err := b.rewriteInsertOnDuplicateUpdate(assign.Expr, mockTablePlan, insertPlan)
-		if err != nil {
-			b.err = errors.Trace(err)
-			return nil
+	// Check that there's no generated column.
```
Comment (zz-jason) [inline plan/planbuilder.go]: @CaitinChen @lilin90 PTAL this comment, thanks

Comment (CaitinChen) [reply plan/planbuilder.go]: Check to guarantee that no column is generated.

File: plan/planbuilder.go (hunk 3 above)
Comment (winoros) [reply plan/planbuilder.go]: Yes, it's for the `else` branch, explaining that else branch doesn't need to do anything. 
I think it's useful. But maybe we can move it other place.

File: plan/logical_plan_test.go [hunk 8]
```diff
@@ -1332,7 +1332,7 @@ func (s *testPlanSuite) TestVisitInfo(c *C) {
 		ans []visitInfo
 	}{
 		{
-			sql: "insert into t values (1)",
```
[the remark this answers is not in the document]
Comment (winoros) [reply plan/logical_plan_test.go]: Yes, this the length of value_list is not same with the number of table's columns.

**Comment by winoros** [comment]
PTAL @zz-jason

File: plan/planbuilder.go [hunk 9]
```diff
@@ -951,154 +935,255 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
-		expr, err := b.rewriteInsertOnDuplicateUpdate(assign.Expr, mockTablePlan, insertPlan)
-		if err != nil {
-			b.err = errors.Trace(err)
-			return nil
+	// Check to guarantee that there's no generated column.
+	// This check is done after the above one is to make compatible with mysql.
+	// i.e. table t have two column a and b where b is generated column.
+	// `insert into t (b) select * from t` will raise a error which tells you that the column count is not matched.
+	// `insert into t select * from t` will raise a error which tells you that there's generated column in column list.
+	// If we do this check before the above one, the first example will raise the error same with the second example.
```
Comment (zz-jason) [inline plan/planbuilder.go]: @CaitinChen Could help us to refine this comment block 😂

File: plan/planbuilder.go [hunk 10]
```diff
@@ -951,154 +935,255 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
...
+		return
+	}
 
-		// Construct the function which calculates the assign value of the column.
-		expr, err := b.rewriteInsertOnDuplicateUpdate(assign.Expr, mockTablePlan, insertPlan)
-		if err != nil {
-			b.err = errors.Trace(err)
-			return nil
+	// Check to guarantee that there's no generated column.
+	// This check is done after the above one is to make compatible with mysql.
```
Comment (CaitinChen) [inline plan/planbuilder.go]: "This check should be done after the above one to make TiDB compatible with MySQL"?

File: plan/planbuilder.go (hunk 3 above)
Comment (zz-jason) [reply plan/planbuilder.go]: I think we can put this comment to the `if len(insert.Columns) > 0 || len(insert.Lists[0]) > 0 {` to explain why we don't check whether there is generated column in the table when the `value_list` and `col_list` is empty.

**Comment by winoros** [comment]
@jackysp PTAL

**Pull request merged**